### ISSUE TITLE
Bug: [Streaming] No reliable way to tell whether error is React internal error or boundary error

### ISSUE THREAD

**Issue by brillout** [facebook/react#24536]
React version: `18.1.0`

## The current behavior

The `onError()` callback of `renderToPipeableStream()` and `renderToReadableStream()` receives boundary errors as well as internal React errors (i.e. React bugs). There isn't a reliable way to tell whether the error is a boundary error or an internal React error.

## The expected behavior

There is a reliable way to tell whether the error is a boundary error or an internal React error.

## Steps To Reproduce

N/A.

Strictly speaking this is not a bug, but I do believe this is an issue. Apologies if this is the wrong place, I'm happy to close and re-open this somewhere else.

(Btw Ricky told me that you are wrapping up the React 18 Working Group, and that I should ask React Streaming questions by creating an issue on React's main repo. I may have more questions about React 18 Streaming.)

## Additional Context

As the author of [`react-streaming`](https://github.com/brillout/react-streaming), I'd like to offer an easy way for React users to consume the React SSR Stream.

Users will handle React internal errors quite differently from handling boundary errors.

A boundary error can safely be ignored since it will be re-tried on the client-side. If the user has error tracking already installed on the client-side, then he may not need to do anything about boundary errors on the server-side.

A React bug is fatal and needs quite some complex consideration to handle properly. I'm expecting most users to just ignore potential React bugs and just assume React to be bug-free. This is not perfect, but handling React bugs is really complex: the stream can be corrupt in all kinds of ways and there is no clear way how to recover from this.

So I expect 95% of the users to ignore potential React bugs.

Bottom line: boundary errors and React bugs will be handled quite differently.

That's why I think there shoud be a reliable way to tell whether `err` in `onError(err)` is a boundary error or a Rect internal error (i.e. a "React bug").

(For `react-streaming` this is low-priority and low-urgency. Fundamentally more critical for `react-streaming` is the RFC about `injectToStream()`.)

**Comment by brillout** [comment]
This is still relevant.

**Comment by brillout** [comment]
From the perspective of `react-streaming` (or from the perspective of React for that matter), user-land errors are expected and should be handled accordingly. Whereas React internal errors should be handled differently (since they are unexpected anything could happen).

For example, most data fetching tools such as `fetch()` and Telefunc throw an error upon network issues. These kind of errors are very much expected. (Essentially, non-expected errors are bugs.)

That said, so far, things are working well so I guess we can close this issue &mdash; we'll see whether the need to distinguish in the future arises. (I don't remember what use case I had in mind when I opened this ticket. Apologies for the noise.)
